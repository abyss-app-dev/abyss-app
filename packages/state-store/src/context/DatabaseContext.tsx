import type { SQliteClient } from '@abyss/records';
import { createContext, type ReactNode, useContext } from 'react';

interface DatabaseContextType {
    database: SQliteClient | null;
}

const DatabaseContext = createContext<DatabaseContextType>({ database: null });

interface DatabaseProviderProps {
    database: SQliteClient;
    children: ReactNode;
}

export function DatabaseProvider({ database, children }: DatabaseProviderProps) {
    return <DatabaseContext.Provider value={{ database }}>{children}</DatabaseContext.Provider>;
}

export function useDatabase(): SQliteClient {
    const context = useContext(DatabaseContext);
    if (!context.database) {
        throw new Error('useDatabase must be used within a DatabaseProvider');
    }
    return context.database;
}
