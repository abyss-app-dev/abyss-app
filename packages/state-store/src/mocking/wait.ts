export function wait(ms = 50) {
    return new Promise(resolve => setTimeout(resolve, ms));
}
