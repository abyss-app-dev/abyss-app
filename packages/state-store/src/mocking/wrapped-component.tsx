import type { StateRecorder } from './state-recorder';

export function TestRecorder<T>({ functionToTest, recorder }: { functionToTest: () => T; recorder: StateRecorder }) {
    recorder.record(functionToTest());
    return <>content</>;
}
