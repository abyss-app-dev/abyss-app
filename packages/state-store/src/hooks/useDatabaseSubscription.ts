import { useEffect } from 'react';
import { useDatabase } from '../context/DatabaseContext';
import type { DbQuery, Dependencies } from './useQuery';
import { useQuery } from './useQuery';

export type UseDatabaseSubscription<T> = ReturnType<typeof useDatabaseSubscription<T>>;

export function useDatabaseSubscription<T>(callback: DbQuery<T>, listeners: Dependencies = []) {
    const database = useDatabase();
    const query = useQuery(callback, listeners);

    useEffect(() => {
        return database.subscribeDatabase(() => query.refetch());
    }, [database, query.refetch]);

    return query;
}
