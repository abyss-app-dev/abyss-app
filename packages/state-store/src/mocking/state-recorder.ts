import { render } from '@testing-library/react';
import type { ReactNode } from 'react';
import { expect } from 'vitest';
import { wait } from './wait';

export class StateRecorder {
    private states: unknown[] = [];

    record(state: unknown): void {
        this.states.push(state);
    }

    async delayedResolve(): Promise<unknown[]> {
        await wait();
        return this.states;
    }

    getStates(): unknown[] {
        return this.states;
    }

    getLatestState(): unknown {
        return this.states[this.states.length - 1];
    }

    clear(): void {
        this.states.length = 0;
    }

    async renderToSnapshot(component: ReactNode): Promise<void> {
        render(component);
        expect(await this.delayedResolve()).toMatchSnapshot();
    }
}
