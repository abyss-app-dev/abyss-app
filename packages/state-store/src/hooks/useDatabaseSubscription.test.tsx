import { buildTestDB, type SQliteClient, SqliteTable } from '@abyss/records';
import { useEffect, useState } from 'react';
import { beforeEach, describe, test } from 'vitest';
import { DatabaseProvider } from '../context/DatabaseContext';
import { StateRecorder } from '../mocking/state-recorder';
import { TestRecorder } from '../mocking/wrapped-component';
import { useDatabaseSubscription } from './useDatabaseSubscription';

describe('useDatabaseSubscription hook', () => {
    let db: SQliteClient;
    let recorder: StateRecorder;

    beforeEach(async () => {
        db = await buildTestDB();
        recorder = new StateRecorder();
    });

    test('basic subscription with database query', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const subscription = useDatabaseSubscription(async database => {
                            return database.tables[SqliteTable.settings].default();
                        });
                        return subscription;
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('subscription with query that returns null', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const subscription = useDatabaseSubscription(async () => {
                            return null;
                        });
                        return subscription;
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('subscription with query that throws error', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const subscription = useDatabaseSubscription(async () => {
                            throw new Error('Database query failed');
                        });
                        return subscription;
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('subscription with dependencies that change', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const [dependency, setDependency] = useState(1);

                        useEffect(() => {
                            if (dependency === 1) {
                                setTimeout(() => setDependency(2));
                            }
                        }, [dependency]);

                        const subscription = useDatabaseSubscription(async () => {
                            return `Query result for dependency: ${dependency}`;
                        }, [dependency]);

                        return subscription;
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('subscription with multiple dependencies', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const [dep1, setDep1] = useState('a');
                        const [dep2, setDep2] = useState(1);

                        useEffect(() => {
                            if (dep1 === 'a' && dep2 === 1) {
                                setTimeout(() => {
                                    setDep1('b');
                                    setDep2(2);
                                });
                            }
                        }, [dep1, dep2]);

                        const subscription = useDatabaseSubscription(async () => {
                            return `Query result for deps: ${dep1}-${dep2}`;
                        }, [dep1, dep2]);

                        return subscription;
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('manually call refetch', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const [hasRefetched, setHasRefetched] = useState(false);

                        const subscription = useDatabaseSubscription(async () => {
                            return `Initial data at ${Date.now()}`;
                        });

                        useEffect(() => {
                            if (subscription.data && !subscription.loading && !hasRefetched) {
                                setHasRefetched(true);
                                setTimeout(() => subscription.refetch());
                            }
                        }, [subscription.data, subscription.loading, subscription.refetch, hasRefetched]);

                        return subscription;
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('manually call setData', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const [hasSetData, setHasSetData] = useState(false);

                        const subscription = useDatabaseSubscription(async () => {
                            return 'initial data';
                        });

                        useEffect(() => {
                            if (subscription.data && !subscription.loading && !hasSetData) {
                                setHasSetData(true);
                                setTimeout(() => subscription.setData('manually set data'));
                            }
                        }, [subscription.data, subscription.loading, subscription.setData, hasSetData]);

                        return subscription;
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('subscription responds to database changes', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const [hasTriggeredChange, setHasTriggeredChange] = useState(false);
                        const [refetchCount, setRefetchCount] = useState(0);

                        const subscription = useDatabaseSubscription(async database => {
                            const records = await database.tables[SqliteTable.settings].list();
                            return `Settings count: ${records.length}`;
                        });

                        // Track refetch calls by monitoring data changes
                        useEffect(() => {
                            if (subscription.data && !subscription.loading) {
                                setRefetchCount(prev => prev + 1);
                            }
                        }, [subscription.data, subscription.loading]);

                        // Trigger a database change that should cause subscription update
                        useEffect(() => {
                            if (subscription.data && !subscription.loading && !hasTriggeredChange) {
                                setHasTriggeredChange(true);
                                setTimeout(async () => {
                                    // Creating a new record should trigger database subscription
                                    await db.tables[SqliteTable.settings].default();
                                }, 100);
                            }
                        }, [subscription.data, subscription.loading, hasTriggeredChange]);

                        return {
                            loading: subscription.loading,
                            error: subscription.error,
                            data: subscription.data,
                            hasTriggeredChange,
                            refetchCount,
                        };
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('subscription with complex database query', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const subscription = useDatabaseSubscription(async database => {
                            const settingsRecords = await database.tables[SqliteTable.settings].list();
                            return {
                                recordCount: settingsRecords.length,
                                firstRecord: settingsRecords[0] || null,
                                timestamp: Date.now(),
                            };
                        });

                        return subscription;
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('transition from data to failure', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const [shouldFail, setShouldFail] = useState(false);

                        const subscription = useDatabaseSubscription(async () => {
                            if (shouldFail) {
                                throw new Error('Query now fails');
                            }
                            return 'successful data';
                        }, [shouldFail]);

                        useEffect(() => {
                            if (subscription.data === 'successful data') {
                                setTimeout(() => setShouldFail(true));
                            }
                        }, [subscription.data]);

                        return subscription;
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('transition from failure to data', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const [shouldFail, setShouldFail] = useState(true);

                        const subscription = useDatabaseSubscription(async () => {
                            if (shouldFail) {
                                throw new Error('Initial failure');
                            }
                            return 'recovered data';
                        }, [shouldFail]);

                        useEffect(() => {
                            if (subscription.error) {
                                setTimeout(() => setShouldFail(false));
                            }
                        }, [subscription.error]);

                        return subscription;
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('subscription with no dependencies uses empty array', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        // Test the default empty array for dependencies
                        const subscription = useDatabaseSubscription(async () => {
                            return 'data with no dependencies';
                        });

                        return subscription;
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('subscription handles database reference changes', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const [queryCount, setQueryCount] = useState(0);

                        const subscription = useDatabaseSubscription(async database => {
                            const newCount = queryCount + 1;
                            setQueryCount(newCount);
                            return `Query executed ${newCount} times with DB: ${!!database}`;
                        });

                        return { ...subscription, queryCount };
                    }}
                />
            </DatabaseProvider>
        );
    });
});
