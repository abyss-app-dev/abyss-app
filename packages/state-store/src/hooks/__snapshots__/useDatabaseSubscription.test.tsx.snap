// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useDatabaseSubscription hook > basic subscription with database query 1`] = `
[
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": {
      "createdAt": 1748759090884,
      "id": "settings::default",
      "lastPage": "/",
      "theme": "",
      "updatedAt": 1748759090884,
    },
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;

exports[`useDatabaseSubscription hook > manually call refetch 1`] = `
[
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "Initial data at 1748759092252",
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "Initial data at 1748759092252",
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "Initial data at 1748759092253",
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;

exports[`useDatabaseSubscription hook > manually call setData 1`] = `
[
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "initial data",
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "initial data",
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "manually set data",
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;

exports[`useDatabaseSubscription hook > subscription handles database reference changes 1`] = `
[
  {
    "data": null,
    "error": null,
    "loading": true,
    "queryCount": 0,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": null,
    "error": null,
    "loading": true,
    "queryCount": 1,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "Query executed 1 times with DB: true",
    "error": null,
    "loading": false,
    "queryCount": 1,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;

exports[`useDatabaseSubscription hook > subscription responds to database changes 1`] = `
[
  {
    "data": null,
    "error": null,
    "hasTriggeredChange": false,
    "loading": true,
    "refetchCount": 0,
  },
  {
    "data": "Settings count: 0",
    "error": null,
    "hasTriggeredChange": false,
    "loading": false,
    "refetchCount": 0,
  },
  {
    "data": "Settings count: 0",
    "error": null,
    "hasTriggeredChange": true,
    "loading": false,
    "refetchCount": 1,
  },
]
`;

exports[`useDatabaseSubscription hook > subscription with complex database query 1`] = `
[
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": {
      "firstRecord": null,
      "recordCount": 0,
      "timestamp": 1748759093038,
    },
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;

exports[`useDatabaseSubscription hook > subscription with dependencies that change 1`] = `
[
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "Query result for dependency: 1",
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "Query result for dependency: 1",
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "Query result for dependency: 1",
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "Query result for dependency: 2",
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;

exports[`useDatabaseSubscription hook > subscription with multiple dependencies 1`] = `
[
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "Query result for deps: a-1",
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "Query result for deps: a-1",
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "Query result for deps: a-1",
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "Query result for deps: b-2",
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;

exports[`useDatabaseSubscription hook > subscription with no dependencies uses empty array 1`] = `
[
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "data with no dependencies",
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;

exports[`useDatabaseSubscription hook > subscription with query that returns null 1`] = `
[
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": null,
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;

exports[`useDatabaseSubscription hook > subscription with query that throws error 1`] = `
[
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": null,
    "error": [Error: Database query failed],
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;

exports[`useDatabaseSubscription hook > transition from data to failure 1`] = `
[
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "successful data",
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "successful data",
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "successful data",
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": null,
    "error": [Error: Query now fails],
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;

exports[`useDatabaseSubscription hook > transition from failure to data 1`] = `
[
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": null,
    "error": [Error: Initial failure],
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": null,
    "error": [Error: Initial failure],
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": null,
    "error": [Error: Initial failure],
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "recovered data",
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;
